Milestone 2: Components and Redux store

shshamim63 opened the pull request:
## :white_check_mark: The Pull Request contains the following task.
___
1. Updated functionality of App component.
2. Generated new component `BookList` that render list of books.
3. Contains a Book component.
4. Added BookForm component that shows the input field.
5. Created actions to create and remove books.
6. A reducer to handle all the dispatched actions.
7. Created redux store and initialized it with predefined value.

macbright committed:
:sparkles: added more new features to the app structure

prepared the structure directory
added the containers directory
added the actions directory
added the reducers directory
created book and index file for the reducer directory
created index filer for the action directory
created a booklist file for the container directory

shshamim63 committed:
:art: Adds new components

Adds BookForm component as a part of containers
Adds BookList component as a part of containers
Changes name of containers.js to containes

macbright committed:
:art: added table to the BookList component

added a JSX code
added the Book ID, Title and Category
linked the booklist component to the redux store
added a html table

shshamim63 committed:
:art: Adds Book component

Adds functional component named Book
Contains a component that return a row of a table with data from props

macbright committed:
:art: added a html form to the bookstore component

added html form
added book title
added book a select box book categories with a predefined constants
added a submit button to the form

shshamim63 committed:
:art: Adds action and reducer

Adds createBook action creation to generate an action
Adds removeBook action
Adds reducer that perform based on the actions

macbright committed:
:art: created the initial state object

initial state object
created a utils file where i declared the initial state object
created a unique id for each book
wrapped the app component with the provider component from react-redux library

shshamim63 committed:
:art: Renders BookList with refactoring code

Display list of book
Fixes linting errors

shshamim63 committed:
:bug: Adds . to the stickler

Adds . before eslintrc.json to fix the path

macbright committed:
:rotating_light: linter error

fix linter error

macbright committed:
:rotating_light: linter error

fix eslint linter error

bolah2009 commented on src/containers/BookForm.js:
```diff
@@ -0,0 +1,34 @@
+          <option value="Learning"> Learning </option>
+          <option value="Sci-Fi"> Sci-Fi </option>
+        </select>
+      </div>
```
Instead of hardcoding this value, let's define a constant of these categories and `map` them into `option` element.

![image](https://user-images.githubusercontent.com/36057474/72048477-276ec500-32bd-11ea-8cd6-98c29fa345a7.png)

bolah2009 commented on src/actions/index.js:
```diff
@@ -0,0 +1,11 @@
+
+export const removeBook = book => ({
+  type: 'REMOVE_BOOK',
+  payload: book,
```
[Optional]

We could return only `bookId` of `book` for the `payload` property instead.

bolah2009 commented on src/components/App.js:
```diff
@@ -1,5 +1,17 @@
+      <BooksList />
+    </div>
+  );
+};
```
Instead of this disabling this, we could refactor the component since in this case, we don't need an implicit `return` keyword.

Something like this could solve the linter warning.

```jsx
const App = () => (
  <div>
    <BookForm />
    <br />
    <BooksList />
  </div>
);
```

bolah2009 commented on src/containers/BookForm.js:
```diff
@@ -0,0 +1,34 @@
+          <option value="Kids"> Kids </option>
+          <option value="Learning"> Learning </option>
+          <option value="Sci-Fi"> Sci-Fi </option>
+        </select>
```
Like the `title` input, we could wrap the `select` element with the `label` tag and define the `htmlFor` and `id` for the `label` and `select` tag respectively. This would help with the two rules disabled `/* eslint-disable no-tabs, jsx-a11y/label-has-for */` and `/* eslint-disable jsx-a11y/label-has-associated-control */` and also make the form more accesible.

bolah2009 commented on src/containers/BookList.js:
```diff
@@ -0,0 +1,32 @@
+);
+BookList.prototype = {
+  books: PropTypes.array,
+};
```
There's a typo with this current PropTypes definition. This is mainly the reason we having linter issues here. I would recommend we remove the linter rule that is been disabled and do the following:

- There seem to be `prototype` been used instead of `propTypes`
- Let's use `arrayOf` to define the type of array we are passing as props
- Since its an array of objects and we are using those object properties directly, let's use `shape` to defined the properties of the object. See https://reactjs.org/docs/typechecking-with-proptypes.html#proptypes for more examples and below for how it should be.


```jsx
BookList.propTypes = {
  books: PropTypes.arrayOf(
    PropTypes.shape({
      bookId: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      category: PropTypes.string.isRequired,
    }).isRequired,
  ).isRequired,
};
```

bolah2009 commented on src/reducers/books.js:
```diff
@@ -0,0 +1,15 @@
+import INITIAL_BOOKS from '../utils/bookObject';
+
+
+const bookList = (state = INITIAL_BOOKS, action) => {
```
[Optional]

Since `createStore` has a `preloadedState` argument, I would recommend we use the `INITIAL_BOOKS` in the `createStore` argument.


See https://redux.js.org/api/createstore/

bolah2009 reviewed (changes requested):
### Awesome job team! 

Good job with the components and reducer setup.

Please check the comments below. Please feel free to ignore comments tagged as `Optional` and question those that are not.


#### :recycle:  Make these changes and submit another code review request. 👍🏾

#### Feel free to reach me on slack if you having issues implementing these features. :relaxed: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

shshamim63 committed:
:ok_hand: Updates the code after getting code review

Uses map to render the category
Fixes typo mistake in prop validation
Adds Id to the action as payload
Updates label with htmlFor along with the id of section

Oghenebrume50 reviewed (approved):
### STATUS: APPROVED

Hi, Team this looks good all changes pointed out by the previous reviewer were made :clap: 
**THIS IS APPROVED** :+1: 

**RAPHAEL NORIODE** :sunglasses:
